storage/http_client.go

```diff
@@ -62,56 +60,12 @@ type httpStorageClient struct {
 func newHTTPStorageClient(ctx context.Context, opts ...storageOption) (storageClient, error) {
 	s := initSettings(opts...)
 	o := s.clientOption
-	config := newStorageConfig(o...)
-
-	var creds *auth.Credentials
-	// In general, it is recommended to use raw.NewService instead of htransport.NewClient
-	// since raw.NewService configures the correct default endpoints when initializing the
-	// internal http client. However, in our case, "NewRangeReader" in reader.go needs to
-	// access the http client directly to make requests, so we create the client manually
-	// here so it can be re-used by both reader.go and raw.NewService. This means we need to
-	// manually configure the default endpoint options on the http client. Furthermore, we
-	// need to account for STORAGE_EMULATOR_HOST override when setting the default endpoints.
-	if host := os.Getenv("STORAGE_EMULATOR_HOST"); host == "" {
-		// Prepend default options to avoid overriding options passed by the user.
-		o = append([]option.ClientOption{option.WithScopes(ScopeFullControl, "https://www.googleapis.com/auth/cloud-platform"), option.WithUserAgent(userAgent)}, o...)
-
-		o = append(o, internaloption.WithDefaultEndpointTemplate("https://storage.UNIVERSE_DOMAIN/storage/v1/"),
-			internaloption.WithDefaultMTLSEndpoint("https://storage.mtls.googleapis.com/storage/v1/"),
-			internaloption.WithDefaultUniverseDomain("googleapis.com"),
-		)
-		// Don't error out here. The user may have passed in their own HTTP
-		// client which does not auth with ADC or other common conventions.
-		c, err := internaloption.AuthCreds(ctx, o)
-		if err == nil {
-			creds = c
-			o = append(o, option.WithAuthCredentials(creds))
-		}
-	} else {
-		var hostURL *url.URL
-
-		if strings.Contains(host, "://") {
-			h, err := url.Parse(host)
-			if err != nil {
-				return nil, err
-			}
-			hostURL = h
-		} else {
-			// Add scheme for user if not supplied in STORAGE_EMULATOR_HOST
-			// URL is only parsed correctly if it has a scheme, so we build it ourselves
-			hostURL = &url.URL{Scheme: "http", Host: host}
-		}
-
-		hostURL.Path = "storage/v1/"
-		endpoint := hostURL.String()
-
-		// Append the emulator host as default endpoint for the user
-		o = append([]option.ClientOption{option.WithoutAuthentication()}, o...)
-
-		o = append(o, internaloption.WithDefaultEndpointTemplate(endpoint))
-		o = append(o, internaloption.WithDefaultMTLSEndpoint(endpoint))
+	defaults, err := defaultHTTPOptions(ctx, o...)
+	if err != nil {
+		return nil, err
```

Comment:
We need to add these options before trying to detect/determine the creds. Same goes for GRPC. Can you move this to the higher level function?